protocol/strict-derivation.md

```diff
@@ -0,0 +1,355 @@
+out later by a backwards-invalidating singular batch derived from a span batch. So if we only
+forwards-invalidate, this becomes easier.
+
+And in any case, a restart or reorg needs to reconstruct the derivation pipeline state from the new
+(sync-start) L1 block, so I don't understand how reconstructing span batch derivation state is
+different from reconstructing the channel bank and batch queue in general. If a restart or reorg
+happens in the middle of deriving singular batches from a channel, we're in a similar situation that
+is solved by going back far enough for the sync-start.
+
+# Strict Batch Ordering
+
+Strict batch ordering requires batches within and across channels to be strictly ordered.
+This removes the necessity to buffer more than one channel, and the channel bank effectively holds
+at most one staging channel, for which it collects frames, that must arrive in order.
+
+The following will explore the open design space, presenting options, and propose a solution.
+In order to guide decisions on open design questions, we will consider the 
+_principle of fastest derivation_: the new rules should lead to faster derivation progress where possible,
+even if it means deriving empty batches earlier. 
+Spoiler: Option 1 will always be the proposed option.
+
+## Out of order frames
+
+There's an open design space around how to handle some scenarios for missing or out of order frames:
+- How do we handle "foreign first frames" -- namely, incoming _first_ frames which do not belong to the current channel?
+  - Option 1: discard the current channel and open a new one for the given frame.
+  - Option 2: discard the foreign frame and wait for the duration of the channel timeout for the
+    next correct frame and only then discard the current staging channel and open a new staging
+    channel for the next first frame.
+- How do we handle the current staging channel if an out of order frame (that is not the first
+frame) arrives? There's only one valid option, to drop the out-of-order frame, keeping the currently
+staged channel. Note that this will give the staging channel a chance to be continued or closed
+properly by the next in-order frame. If this doesn't happen, at the latest the next first frame
+will drop the staging channel and start a new one.
+
+### Proposed solution
+
+Following the principle of fastest derivation, this means choosing option 1. This has the
+additional advantage that the sync start algorithm simplifies significantly, because a fist frame
+_always_ marks the start of a new channel, and guarantees that there won't be any buffered channels
+in the channel bank.
+
+## Out of order batches
+
+The batch queue also becomes simpler as batches have to arrive in order. 
+
+- How to treat _older_ batches is easy: they are dropped.
+- Similarly to out of order frames, how to treat future batches (gaps) has two options:
+  - Option 1: immediately fill the gap with deposit-only blocks, and then process the batch.
```

Comment:
I agree with option 1, but an implementation detail we need to keep an eye on is that we don't run out of memory, and stream out the deposit-only blocks that fill the gap, rather than creating them all at once. Since the gap can be very large.

And the gap might have to be limited in size for fault-proofs still. Even with sufficient memory, and streaming rather than all-at-once allocation, it might still require a lot of processing.
However, without making the channel-window the effective new smaller sequencing window, there might still be a lot of deposit-only blocks to churn through.

Reply:
An implementation hint about this should be added to the specs